### PR TITLE
Homography example with functorch

### PR DESCRIPTION
**Pull request by fantaosha**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by ddetone** [comment]
Taking a look now.

High-level comment - Its a little hard to see the diff versus the version we wrote since the base branch is main and not experimental

**Comment by ddetone** [comment]
I tried running this on my cpu machine and got the following error: 
```
> python3 homography_estimation.py
Found 323 total images in dataset
Using 290 images for training
Found 323 total images in dataset
Using 33 images for testing
Writing to tensorboard at /Users/ddetone/code/theseus/examples/outputs/2022-06-28/17-13-07
Using device cpu
Pool size 53, error dim 9

 Starting training for 10 epochs
/Users/ddetone/miniconda3/lib/python3.9/site-packages/torch/functional.py:568: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  /Users/distiller/project/conda/conda-bld/pytorch_1646756029501/work/aten/src/ATen/native/TensorShape.cpp:2228.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
Error executing job with overrides: []
Traceback (most recent call last):
  File "/Users/ddetone/code/theseus/examples/homography_estimation.py", line 824, in main
    run(cfg)
  File "/Users/ddetone/code/theseus/examples/homography_estimation.py", line 759, in run
    net_feat = train_loop(
  File "/Users/ddetone/code/theseus/examples/homography_estimation.py", line 583, in train_loop
    layer_feat.forward(
  File "/Users/ddetone/code/theseus/theseus/theseus_layer.py", line 68, in forward
    vars, info = _forward(
  File "/Users/ddetone/code/theseus/theseus/theseus_layer.py", line 128, in _forward
    info = optimizer.optimize(**optimizer_kwargs)
  File "/Users/ddetone/code/theseus/theseus/optimizer/optimizer.py", line 39, in optimize
    return self._optimize_impl(**kwargs)
  File "/Users/ddetone/code/theseus/theseus/optimizer/nonlinear/nonlinear_optimizer.py", line 349, in _optimize_impl
    self._optimize_loop(
  File "/Users/ddetone/code/theseus/theseus/optimizer/nonlinear/nonlinear_optimizer.py", line 225, in _optimize_loop
    self.linear_solver.linearization.linearize()
  File "/Users/ddetone/code/theseus/theseus/optimizer/linearization.py", line 54, in linearize
    self.objective.update_vectorization()
  File "/Users/ddetone/code/theseus/theseus/core/objective.py", line 486, in update_vectorization
    self._vectorization_run()
  File "/Users/ddetone/code/theseus/theseus/core/vectorizer.py", line 305, in _vectorize
    Vectorize._compute_error_and_replace_wrapper_caches(
  File "/Users/ddetone/code/theseus/theseus/core/vectorizer.py", line 269, in _compute_error_and_replace_wrapper_caches
    v_jac, v_err = vectorized_cost_fn.weighted_jacobians_error()
  File "/Users/ddetone/code/theseus/theseus/core/cost_function.py", line 65, in weighted_jacobians_error
    jacobian, err = self.jacobians()
  File "/Users/ddetone/code/theseus/examples/homography_estimation.py", line 294, in jacobians
    jacobians = [vmap_helper_jac_fn(
  File "/Users/ddetone/miniconda3/lib/python3.9/site-packages/functorch/_src/vmap.py", line 383, in wrapped
    batched_outputs = func(*batched_inputs, **kwargs)
  File "/Users/ddetone/code/theseus/examples/homography_estimation.py", line 291, in helper_jac_fn_functorch
    return jacrev(helper_err_fn)(H)
  File "/Users/ddetone/miniconda3/lib/python3.9/site-packages/functorch/_src/eager_transforms.py", line 419, in wrapper_fn
    vjp_out = vjp(f_wrapper, *primals, has_aux=has_aux)
  File "/Users/ddetone/miniconda3/lib/python3.9/site-packages/functorch/_src/eager_transforms.py", line 248, in vjp
    primals_out = func(*diff_primals)
  File "/Users/ddetone/miniconda3/lib/python3.9/site-packages/functorch/_src/eager_transforms.py", line 602, in f_wrapper
    return f(*replaced_args)
  File "/Users/ddetone/code/theseus/examples/homography_estimation.py", line 285, in helper_err_fn
    return homography_error_fn(
  File "/Users/ddetone/code/theseus/examples/homography_estimation.py", line 244, in homography_error_fn
    loss = (loss * mask).mean(dim=1)
RuntimeError: The size of tensor a (4096000) must match the size of tensor b (200) at non-singleton dimension 4

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

**Comment by ddetone** [comment]
@fantaosha  have you done any basic benchmarking such as GPU memory consumption before / after this addition?

File: examples/homography_estimation.py [hunk 1]
```diff
@@ -0,0 +1,828 @@
...
+    img_init = torch.zeros(batch_size, channels, cfg.imgH, cfg.imgW)
+    img1 = th.Variable(data=img_init, name="img1")
+    img2 = th.Variable(data=img_init, name="img2")
+
+    # loss is pooled so error dim is not too large
+    pool_size = int(cfg.imgH // 3)
+    pooled = torch.nn.functional.avg_pool2d(img_init, pool_size, stride=pool_size)
+    error_dim = pooled[0, 0].numel()
+    print(f"Pool size {pool_size}, error dim {error_dim}")
+    use_functorch = True
```
Comment (ddetone) [inline examples/homography_estimation.py:491]: Make this a parameter in hydra?

**Comment by ddetone** [comment]
I just realized @fantaosha, that I think you made this change on an outdated version of the homography example.

Can you instead base your changes on this version of the homography example? https://github.com/facebookresearch/theseus/blob/experimental/examples/homography_estimation.py

**Comment by fantaosha** [comment]
Will create another PR using `taoshaf.homography_functorch`

**Pull request closed without merging**